src/main/java/com/mongodb/spark/sql/connector/read/partitioner/PaginatePartitioner.java

```diff
@@ -108,6 +107,7 @@ private List<BsonDocument> createUpperBounds(
         boundaryPipeline.addAll(aggregationPipeline);
         boundaryPipeline.add(Aggregates.skip(numDocumentsPerPartition));
         boundaryPipeline.add(Aggregates.limit(1));
+        boundaryPipeline.add(Aggregates.project(projection));
```

Comment:
Should we add a comment to avoid mistakenly re-ordering the aggregation? something like 
 
```suggestion
        // Move projection after skip/limit to reduce memory usage
        // and improve performance on large documents
        boundaryPipeline.add(Aggregates.project(projection));
```